content/blog/2023/11/2023-11-29-jenkins-to-github.adoc

```diff
@@ -0,0 +1,270 @@
+You'll have to directly interact with the GitHub API without the Jenkins layer in such cases.
+In terms of the fancy plugins, this means you even don't have to use the https://github.com/jenkinsci/checks-api-plugin[Checks API Plugin].
+
+For GitHub Apps, the authorization & authentication process can be quite involved:
+
+* Back when you installed your GitHub App to your repo, you granted a set of permissions to your App.
+This is called *authorization*, and exactly what your App can do on your repo can be retrieved from GitHub via an "Installation ID".
+* When your script initializes, it needs to authenticate to GitHub as your App.
+To do so, the script needs a private key of the App.
+You can generate a private key from the App's settings page and use it forever.
+* If dealing with the raw GitHub API, the authentication is quite involved: You first generate a https://jwt.io/introduction[_JSON Web Token_ (JWT)] with the private key.
+Then, you https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation[call a GitHub endpoint] with this JWT, saying "I'm this App, and I'm trying to access the repos we agreed upon earlier." This endpoint gives you an "installation access token" (IAT).
+Your script would use this IAT to call other endpoints, just like how you'd use a PAT when calling endpoints manually as yourself.
+
+Luckily, many languages have libraries that handles the intermediate work for you.
+Let's say your Jenkins file runs a Python script via (`sh './run.py'`).
+We can use the https://github.com/PyGithub/PyGithub[PyGitHub] library to call GitHub APIs Pythonically.
+
+In Python, the built-in https://docs.python.org/3/library/logging.html[logging] library is the standard way to emit messages.
+Your script would create a "logger" and attaches different "handlers" to it, each defining an output channel of your messages -- standard output, file on disk, remote service, etc.
+Intuitively, we can write a handler for logging to GitHub Checks.
+
+With the help of ChatGPT, I implemented one https://gist.github.com/tslmy/84f34a25babe045eb302ec72f2bf39eb[here].
+You can see that the handler requires quite some parameters to initialize.
+Where do you get them?
+
+* For locating the PR and the commit, the https://plugins.jenkins.io/github-pullrequest/[GitHub Integration Plugin] can help.
+It can trigger Jenkins jobs upon PR events, injecting https://github.com/KostyaSha/github-integration-plugin/blob/master/docs/Configuration.adoc#available-environment-variables[these environment variables] to each Build.
+We'll use two: `GITHUB_PR_HEAD_SHA` and `GITHUB_REPO_SSH_URL`.
+* For referencing from the PR Check to the Jenkins Build, we'll also need an env.
+var.
+set by vanilla Jenkins: `BUILD_URL`.
+You can find a list of all Jenkins-set variables https://wiki.jenkins.io/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-belowJenkinsSetEnvironmentVariables[here].
```

Comment:
```suggestion
On link:https://wiki.jenkins.io/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-belowJenkinsSetEnvironmentVariables[this page], you can find a list of all Jenkins-set variables.
```
Suggestion to rearrange the link & syntax so that it is not using the [here] link text.